effects/videoeffect.h

```diff
@@ -19,29 +19,37 @@ extern "C"
 class MP4PlaybackEffect : public LEDEffectBase
 {
 private:
-    string _filePath;
-    AVFormatContext* _formatCtx = nullptr;
-    AVCodecContext* _codecCtx = nullptr;
-    AVFrame* _frame = nullptr;
-    AVPacket* _packet = nullptr;
-    SwsContext* _swsCtx = nullptr;
-    int _videoStreamIndex = -1;
+    string                  _filePath;
+    AVFormatContext*        _formatCtx = nullptr;
+    AVCodecContext*         _codecCtx = nullptr;
+    AVFrame*                _frame = nullptr;
+    AVPacket*               _packet = nullptr;
+    SwsContext*             _swsCtx = nullptr;
+    int                     _videoStreamIndex = -1;
+    atomic<bool>            _initialized = false;
+    mutable recursive_mutex _ffmpegMutex; // recursive_mutex to allow multiple locks by the same thread
+
+    bool AfterInitError(const string& message)
+    {
+        logger->error(message);
+        CleanupFFmpeg();
+        return false;
+    }
```

Comment:
I was extremely tempted to make this a macro, but managed to resist the urge. :)

Reply:
I miss Windows NT exception handling, which had a __finally() handler.  So you'd do all of this code and then just CleanupFFmpeg() in the __finally() handler.